server/telemetry.go

```diff
@@ -68,3 +94,53 @@ func (p *Plugin) initTelemetry(enableDiagnostics *bool) error {
 	}
 	return nil
 }
+
+func (p *Plugin) handleTrackEvent(w http.ResponseWriter, r *http.Request) {
+	var res httpResponse
+	defer p.httpAudit("handleTrackEvent", &res, w, r)
+
+	p.mut.RLock()
+	telemetryEnabled := p.telemetry != nil
+	p.mut.RUnlock()
+
+	if !telemetryEnabled {
+		res.Err = "telemetry is disabled"
+		res.Code = http.StatusBadRequest
+		return
+	}
+
+	var data trackEventRequest
+	if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, requestBodyMaxSizeBytes)).Decode(&data); err != nil {
+		res.Err = err.Error()
+		res.Code = http.StatusBadRequest
+		return
+	}
+
+	if _, ok := telemetryClientEvents[data.Event]; !ok {
+		res.Err = "invalid telemetry event"
+		res.Code = http.StatusBadRequest
+		return
+	}
+
+	if _, ok := telemetryClientTypes[data.ClientType]; !ok {
+		res.Err = "invalid client type"
+		res.Code = http.StatusBadRequest
+		return
+	}
+
+	if data.Props == nil {
+		data.Props = map[string]interface{}{}
+	}
+
+	if data.Source != "" {
+		data.Props["Source"] = data.Source
+	}
+
+	data.Props["ActualUserID"] = r.Header.Get("Mattermost-User-Id")
```

Comment:
Just for my own interest, is this what we're supposed to do? (In terms of privacy)

Reply:
Cannot speak in terms of privacy but we've been adding user ids since forever. Every time you switch a team (or do any other tracked action) you are sending a `user_actual_id` in the payload. Generally we are treating ids as fully opaque values, in fact they are not considered a concern from a purely security perspective.

In this specific case I find it important to track since it's the only piece of authenticated information. Everything else is coming from the client and potentially untrustable.